### DISCUSSION TITLE
Insert a lot of rows using asyncpg driver leads to event loop blocking

### DISCUSSION THREAD

**Discussion by alvassin** [sqlalchemy/sqlalchemy#10070] [Usage Questions]
I need to insert a lot of rows to the database in async application.

Previously i was using asyncpgsa, and there was a problem: it rendered insert statement with all parameters at one time, even if there was a lot of rows, e.g.: `INSERT INTO table (col) VALUES ($1::int), ($2::int), ..., ($32000::int)`, which could led to event loop blocking for really big queries (and parameter limit overflow exceptions).

When i switched to SQLAlchemy, I found that it uses postgres extended query protocol much better: it sends one parse command `INSERT INTO table (col) VALUES ($1::int)`, and then uses `bind` and `execute` commands for each row.

But anyway i see event loop blocking warnings when inserting a lot of rows. How I can fix event loop blocking? 

Please find example below:

```
import asyncio

from sqlalchemy import MetaData, Table, Column, String, Integer, bindparam
from sqlalchemy.ext.asyncio import create_async_engine

values = [{'id': i, 'name': str(i)} for i in range(0, 100_000)]

async def main():
    metadata = MetaData()
    example_table = Table(
        'example', 
        metadata,
        Column('id', Integer, primary_key=True),
        Column('name', String)
    )

    engine = create_async_engine('postgresql+asyncpg://pg:hackme@localhost/pg')
    async with engine.begin() as conn:
        await conn.run_sync(metadata.create_all)
        query = example_table.insert().values(id=bindparam('id'), name=bindparam('name'))
        await conn.execute(query, values)

asyncio.run(main(), debug=True)
```

**Comment by zzzeek** [comment]
hey there- 

**edit** - see my answer below.  I hadn't seen Python do this warning before

I have absolutely no idea (well I have some ideas) what asyncpg is doing here.    It appears to print some messages that I've never seen before (Edit: or is this Python 3.11 trying to tell me that a task is CPU bound?  have never seen this happen):

```
Executing <Task pending name='Task-1' coro=<main() running at /mnt/photon_home/classic/dev/sqlalchemy/test3.py:18> wait_for=<Future pending cb=[Task.task_wakeup()] created at /usr/lib64/python3.11/asyncio/base_events.py:427> cb=[_run_until_complete_cb() at /usr/lib64/python3.11/asyncio/base_events.py:180] created at /usr/lib64/python3.11/asyncio/runners.py:100> took 0.126 seconds
Executing <Task pending name='Task-1' coro=<main() running at /mnt/photon_home/classic/dev/sqlalchemy/test3.py:22> wait_for=<Future pending cb=[Protocol._on_waiter_completed(), Task.task_wakeup()] created at /usr/lib64/python3.11/asyncio/base_events.py:427> cb=[_run_until_complete_cb() at /usr/lib64/python3.11/asyncio/base_events.py:180] created at /usr/lib64/python3.11/asyncio/runners.py:100> took 0.578 seconds

```

in any case, you can lose the .values() portion of your statement, it's automatic

     query = example_table.insert()

seems to reduce this strange executor batching to just one message

```
Executing <Task pending name='Task-1' coro=<main() running at /mnt/photon_home/classic/dev/sqlalchemy/test3.py:22> wait_for=<Future pending cb=[Protocol._on_waiter_completed(), Task.task_wakeup()] created at /usr/lib64/python3.11/asyncio/base_events.py:427> cb=[_run_until_complete_cb() at /usr/lib64/python3.11/asyncio/base_events.py:180] created at /usr/lib64/python3.11/asyncio/runners.py:100> took 0.312 seconds
```

**Reply by zzzeek** [reply]
cc @CaselIT what is this doing?

**Reply by CaselIT** [reply]
This is definitely not the insertmanyrows of sqlalchemy doing this, right?

Does psycopg have the same behaviour?

**Reply by zzzeek** [reply]
it looks like asyncio does this for some flavors of CPU blocking: https://stackoverflow.com/questions/70455957/quart-framework-warningasyncioexecuting

**Reply by 2Prachita** [reply]
This worked for me. I was not getting a straightforward error and I keep going round and round trying to insert the data.
Thank you!

**Comment by zzzeek** [comment]
OK it looks like asyncio does this starting here: https://github.com/python/cpython/blob/e7cd55753b00d6f7e5c8998bae73ebaa9e86398d/Lib/asyncio/base_events.py#L405 and then at https://github.com/python/cpython/blob/e7cd55753b00d6f7e5c8998bae73ebaa9e86398d/Lib/asyncio/base_events.py#L1945 sends out this warning.

The warning seems to be if you have enough rows at once, just organizing the dictionaries into tuples will cause this warning.

There's nothing you can do other than batch + sleep:

```py
        while values:
            batch = values[0:10000]
            values[:10000] = []
            await conn.execute(query, batch)

            await asyncio.sleep(.001)

```
